### PR TITLE
Fixed missing audit events after navigating to settings

### LINKED ISSUE #5551: Missing event in Track Changes audit log when the user exits the form by going to project settings
#### ODK Collect version
2022.4.4, the store version 2023.1.2, the master version 1f9155c7f1f6c1d91532993fd45cc9b68270124a
#### Android version
10, 11, 13
#### Device used
Redmi 9T, Galaxy M12, Galaxy A32
#### Problem description
In a Track Changes form when the user exits the form by going to project settings (and e.g. changes the theme) and later on saves the form, the question on which the user exited the form is missing in the audit log. In View Sent Form all the answers are present. The same issue appears when the user kills the app instead of going to project settings.
#### Steps to reproduce the problem
1. Go to Track Changes Reason 2q form.
[track-changes-reason-on-edit-2q.xml.txt](https://github.com/getodk/collect/files/11232991/track-changes-reason-on-edit-2q.xml.txt)
2. Fill the first and the second question.
3. While on the second question page, go to the Project setting and change the theme of the app.
4. Go to Fill Blank Form and reopen the form.
5. While on the hierarchy view, tap "Go to end”.
6. Finalize and save the form.
7. Enter and save the reason for changes.
8. Go to audit log file and check if both answers are present in the audit log.
#### Expected behavior
All answers should be in the audit log.

### PR DESCRIPTION
**Pull request by grzesiek2010**
Closes #5551 

#### What has been done to verify that this works as intended?
I've added automated tests and tested the fix manually.

#### Why is this the best possible solution? Were any other approaches considered?
The problem was that after going to settings (which is a separate activity) answers from the current screen weren't saved and audit events related to those answers weren't saved as well. I think we should just do that before navigating to settings.

#### How does this change affect users? Describe intentional changes to behavior and behavior that could have accidentally been affected by code changes. In other words, what are the regression risks?
Generally, the fix should be safe and it should only address the issue but you know how complicated logging events is... 

#### Do we need any specific form for testing your changes? If so, please attach one.
No.

#### Does this change require updates to documentation? If so, please file an issue [here]( https://github.com/getodk/docs/issues/new) and include the link below.
No.

#### Before submitting this PR, please make sure you have:
- [x] run `./gradlew checkAll` and confirmed all checks still pass OR confirm CircleCI build passes and run `./gradlew connectedDebugAndroidTest` locally.
- [x] verified that any code or assets from external sources are properly credited in comments and/or in the [about file](https://github.com/getodk/collect/blob/master/collect_app/src/main/assets/open_source_licenses.html).
- [x] verified that any new UI elements use theme colors. [UI Components Style guidelines](https://github.com/getodk/collect/blob/master/docs/CODE-GUIDELINES.md#ui-components-style-guidelines)

### CODE REVIEW

File: collect_app/src/main/java/org/odk/collect/android/formentry/FormEntryMenuDelegate.java [hunk 1]
```diff
@@ -111,6 +111,7 @@ public boolean onOptionsItemSelected(MenuItem item) {
             if (audioRecorder.isRecording()) {
                 DialogFragmentUtils.showIfNotShowing(RecordingWarningDialogFragment.class, activity.getSupportFragmentManager());
             } else {
+                formEntryViewModel.updateAnswersForScreen(answersProvider.getAnswers(), false);
```
Comment (seadowg) [inline collect_app/src/main/java/org/odk/collect/android/formentry/FormEntryMenuDelegate.java:114]: Is it possible we should just be updating answers in `onPause` (instead of in `onSaveInstanceState`) to capture all these kinds of cases? It feels like we want to be saving answers whenever the user "leaves" form entry. Can we give that a try?

**Review by seadowg — CHANGES_REQUESTED** [review]

File: collect_app/src/main/java/org/odk/collect/android/formentry/FormEntryMenuDelegate.java (hunk 1 above)
Comment (grzesiek2010) [reply collect_app/src/main/java/org/odk/collect/android/formentry/FormEntryMenuDelegate.java:114]: I was thinking about this and I think it was there in the past but you got rid of it in one of your prs. I'm pretty sure that while solving this issue I even found that pull request but now I can't recall what pr it was... or maybe I just had a dream :smile:

Comment (seadowg) [reply collect_app/src/main/java/org/odk/collect/android/formentry/FormEntryMenuDelegate.java:114]: > I was thinking about this and I think it was there in the past but you got rid of it in one of your prs.

That's probably right! I think it would be good to investigate using `onPause` instead `onSaveInstanceState`. The difference now is that we have more ways to test Activity lifecycle, save points and audit logs so I think we could write a test that checks the log at `onPause` if that is a better way to do it.

Comment (grzesiek2010) [reply collect_app/src/main/java/org/odk/collect/android/formentry/FormEntryMenuDelegate.java:114]: Ah it was this chaneg https://github.com/getodk/collect/commit/bdcb462dff73564188f4634f7875a0e82491ad24
Do you remember the details? If so do you still think we should move that back to `onPause`?

Comment (seadowg) [reply collect_app/src/main/java/org/odk/collect/android/formentry/FormEntryMenuDelegate.java:114]: I guess it must have been contributing to https://github.com/getodk/collect/issues/5291 or https://github.com/getodk/collect/issues/5301. I'd be interested to see if adding it back in again causes either issue to reappear. And again, our ability to test this kind of stuff is in a way better place so it should be more viable to deal with any edge cases.

Comment (seadowg) [reply collect_app/src/main/java/org/odk/collect/android/formentry/FormEntryMenuDelegate.java:114]: And just for additional context on my thinking here: I do think it's worth doing bigger reworks here now rather than just do the minimal changes to fix the bugs now we have the ability to write better tests.

**Review by seadowg — CHANGES_REQUESTED** [review]
See inline comment

File: collect_app/src/main/java/org/odk/collect/android/formentry/FormEntryMenuDelegate.java (hunk 1 above)
Comment (grzesiek2010) [reply collect_app/src/main/java/org/odk/collect/android/formentry/FormEntryMenuDelegate.java:114]: Ok I've investigated using `onPause` or any other method called when an activity is being closed/recreated like `onStop`/`onDestroy` and I'm rather sure we can't use any of those to solve this issue and make the code clearer.
To solve this issue we want to update answers and flush the audit logger to make sure the answer is saved when we navigate to settings. However, we don't want to flush the logger in `onPause`/`onStop`/`onDestroy` because this would save logs for every screen rotation for example. Basically, this would do more harm than good.
If only we had a way to distinguish between activity recreation (when a device is rotated) and real form closing when the activity is killed then we could implement some improvements but in Android it's not easy.

Comment (seadowg) [reply collect_app/src/main/java/org/odk/collect/android/formentry/FormEntryMenuDelegate.java:114]: > However, we don't want to flush the logger in onPause/onStop/onDestroy because this would save logs for every screen rotation for example.


Ah of course!

**Review by seadowg — APPROVED** [review]

**Comment by dbemke** [comment]
**Tested with Success!**

Verified on device with Android 10

**Verified cases:**
- issue #5551 is no longer reproducing
- regression tests in forms with audit

**Comment by srujner** [comment]
**Tested with Success!**

Verified on device with Android 13
